kernel/src/hil/block_storage.rs

```diff
@@ -0,0 +1,247 @@
+    pub fn new_starting_at(address: u64) -> Option<Self> {
+        if address % S as u64 == 0 {
+            Some(Self::new_containing(address))
+        } else {
+            None
+        }
+    }
+}
+
+impl<const S: usize> From<BlockIndex<S>> for u64 {
+    fn from(index: BlockIndex<S>) -> Self {
+        index.0 as u64 * S as u64
+    }
+}
+
+impl<const S: usize> Add<u32> for BlockIndex<S> {
+    type Output = Self;
+    fn add(self, other: u32) -> Self {
+        BlockIndex(self.0 + other)
+    }
+}
+
+/// Readable persistent block storage device.
+///
+/// The device is formed from equally-sized storage blocks,
+/// which are arranged one after another, without gaps or overlaps,
+/// to form a linear storage of bytes.
+///
+/// Every byte on the device belongs to exactly one block.
+///
+/// `R`: The size of a read block in bytes.
+pub trait ReadableStorage<const R: usize> {
+    /// Read data from a block, and into the buffer.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `region` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// Returns `ErrorCode::BUSY` when another operation is in progress.
+    ///
+    /// On success, triggers `ReadableClient::read_complete` once.
+    fn read(
+        &self,
+        region: &BlockIndex<R>,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Returns the size of the device in bytes.
+    fn get_size(&self) -> u64;
+}
+
+/// Writeable persistent block storage device.
+///
+/// The device is formed from equally-sized storage blocks,
+/// which are arranged one after another, without gaps or overlaps,
+/// to form a linear storage of bytes.
+///
+/// The device is split into blocks in two ways, into:
+/// - discard blocks, which are the smallest unit of space
+/// that can be discarded (see `BlockStorage::discard`)
+/// - write blocks, which are the smallest unit of space that can be written
+///
+/// Every byte on the device belongs to exactly one discard block,
+/// and to exactly one write block at the same time.
+///
+/// `W`: The size of a write block in bytes.
+/// `D`: The size of a discard block in bytes.
+pub trait WriteableStorage<const W: usize, const D: usize> {
+    /// Write data from a buffer to storage.
+    ///
+    /// This function writes the contents of `buf` to memory,
+    /// at the chosen `region`.
+    ///
+    /// `ErrorCode::INVAL` will be returned if
+    /// - `region` exceeds the end of the device, or
+    /// - `buf` is shorter than `region`.
+    ///
+    /// This function SHALL NOT discard the block first.
+    /// The user of this function MUST ensure that the relevant block
+    /// has been successfully discarded first (see `discard`).
+    ///
+    /// Once a byte has been written as part of a write block,
+    /// it MUST NOT be written again until it's discarded
+    /// as part of a discard block.
```

Comment:
This is very restrictive and doesn't match a lot of flash use cases

Reply:
Covering all use cases is not the goal at this stage, see discussion leading up to https://github.com/tock/tock/pull/2993#issuecomment-1207910434

Reply:
I would say allowing multiple writes doesn't mean covering all use cases, just a large number of them

Reply:
I agree, but I still think it's better to tackle this as a separate unit of work. Incremental progress seems to be working out better in Tock than general solutions in my experience.

Reply:
I agree -- all of previous attempts to write a powerful but general flash API have failed because of all of the tricky edge cases: the APIs worked onlu for the particular devices/use cases the author cared about. We can go down that road again, but this shouldn't block having a simpler API for simpler use cases.